src/expr/src/linear.rs

```diff
@@ -372,7 +374,17 @@ impl MapFilterProject {
             available_expr[*index] = true;
         }
         for expr in self.expressions.into_iter() {
-            let is_available = expr.support().into_iter().all(|i| available_expr[i]);
+            // We treat an expression as available if its supporting columns are available,
+            // and if it is not a literal (we want to avoid pushing down literals). This
+            // choice is ad-hoc, but the intent is that we partition the operators so
+            // that we can reduce the row representation size and total computation.
+            // Pushing down literals harms the former and does nothing for the latter.
+            // In the future, we'll want to have a harder think about this trade-off, as
+            // we are certainly making sub-optimal decisions by pushing down all availble
+            // work.
+            // TODO(mcsherry): establish better principles about what work to push down.
```

Comment:
Comment that is not necessarily actionable:
 
I think better principles include:
* Is the column being mapped demanded? If the column being mapped cannot be eliminated after the map filter project, then the map is just producing unnecessary columns.
* What type is the column being mapped to/from? If an int is being mapped to a string, it is probably better to leave it as an int.

Reply:
These are good points, and we should probably try and capture them in notes somewhere. I'll drop them in to #query-optimization for now, but I can see a few upcoming moments where we'll have the chance again.